### PR TITLE
Abstract webhook logic to prepare for rotating certificates

### PR DESCRIPTION
**Pull request by steeling**
Abstract the webhook logic so that it shares a common base set of methods for creating the server.

This is in preparation to allow these webhook servers to rotate their certificates

There's a couple things this PR does:

1. Leverage a single, underlying webhook server that starts up the appropriate server with TLS and an ability to rotate certificates.
2. remove ports as flags as these were neither exposed through helm, and there was a bug where if a user *did* specify a different port via the flags, it wouldn't actually be plumbed through properly.
3. Remove inconsistency with health checks across the 3 backend systems, by creating a simple healthcheck on the metrics port.
4. Switch the TLS Config from a hard coded cert to a getter function, to allow for *future* rotation

Part of #4839

### CODE REVIEW

File: cmd/osm-injector/osm-injector.go [hunk 1]
```diff
@@ -241,6 +242,7 @@ func main() {
 	}
 
 	<-stop
+	cancel()
```
Comment (nojnhuh) [inline cmd/osm-injector/osm-injector.go:245]: Is there a reason this is necessary on top of the `defer cancel()` where this is created? And it looks like this was already being called at more-or-less this same point in time from within `signals.RegisterExitHandlers`.

File: pkg/webhook/server.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	// set the certificate once.
+	if err := s.setCert(name, namespace); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+// Run actually starts the server. It blocks until the passed in context is done.
+func (s *Server) Run(ctx context.Context) {
+	log.Info().Msgf("Starting conversion webhook server on: %s", s.server.Addr)
```
Comment (jaellio) [inline pkg/webhook/server.go:62]: nit: similar to Jon's comment below 
```suggestion
	log.Info().Msgf("Starting webhook server on: %s", s.server.Addr)
```

File: pkg/injector/webhook.go [hunk 3]
```diff
@@ -39,103 +37,42 @@ const (
...
-	w.WriteHeader(http.StatusOK)
-	if _, err := w.Write([]byte("Health OK")); err != nil {
-		log.Error().Err(err).Msg("Error writing bytes for mutating webhook health check handler")
+	// Start the MutatingWebhook web server
+	srv, err := webhook.NewServer(constants.OSMInjectorName, osmNamespace, constants.InjectorWebhookPort, certManager, map[string]http.HandlerFunc{
+		webhookCreatePod: http.HandlerFunc(wh.podCreationHandler),
+	},
+		func(cert *certificate.Certificate) error {
+			if err := createOrUpdateMutatingWebhook(kubeClient, cert, webhookTimeout, webhookConfigName, meshName, osmNamespace, osmVersion, enableReconciler); err != nil {
+				return errors.Errorf("Error creating MutatingWebhookConfiguration %s: %+v", webhookConfigName, err)
```
Comment (jaellio) [inline pkg/injector/webhook.go:67]:
```suggestion
				return errors.Errorf("Error creating MutatingWebhookConfiguration %s: %+v", webhookConfigName, err)
```
```suggestion
				return errors.Errorf("Error creating or updating MutatingWebhookConfiguration %s: %+v", webhookConfigName, err)
```

File: pkg/validator/server.go [hunk 4]
```diff
@@ -58,12 +47,20 @@ func NewValidatingWebhook(webhookConfigName, osmNamespace, osmVersion, meshName
...
 	}
 
-	// Create the ValidatingWebhook
-	if err := createOrUpdateValidatingWebhook(kubeClient, webhookHandlerCert, webhookConfigName, meshName, osmNamespace, osmVersion, validateTrafficTarget, enableReconciler); err != nil {
-		return errors.Errorf("Error creating ValidatingWebhookConfiguration %s: %+v", webhookConfigName, err)
+	srv, err := webhook.NewServer(ValidatorWebhookSvc, osmNamespace, constants.ValidatorWebhookPort, certManager, map[string]http.HandlerFunc{
+		validationAPIPath: v.doValidation,
+	}, func(cert *certificate.Certificate) error {
+		if err := createOrUpdateValidatingWebhook(kubeClient, cert, webhookConfigName, meshName, osmNamespace, osmVersion, validateTrafficTarget, enableReconciler); err != nil {
+			return errors.Errorf("Error creating ValidatingWebhookConfiguration %s: %+v", webhookConfigName, err)
```
Comment (jaellio) [inline pkg/validator/server.go:54]:
```suggestion
			return errors.Errorf("Error creating or updating ValidatingWebhookConfiguration %s: %+v", webhookConfigName, err)
```

File: pkg/webhook/server.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+		TLSConfig: &tls.Config{
+			GetCertificate: func(chi *tls.ClientHelloInfo) (*tls.Certificate, error) {
+				s.mu.Lock()
+				defer s.mu.Unlock()
+				return &s.cert, nil
+			},
+			MinVersion: tls.VersionTLS13,
+		},
+	}
+	// set the certificate once.
```
Comment (jaellio) [inline pkg/webhook/server.go:53]: Maybe this comment could clarify the webhook configuration is also being created.

File: cmd/osm-injector/osm-injector.go (hunk 1 above)
Comment (steeling) [reply cmd/osm-injector/osm-injector.go]: nope, removed that call

File: pkg/validator/server_test.go [hunk 6]
```diff
@@ -162,15 +162,14 @@ func TestNewValidatingWebhook(t *testing.T) {
 		tassert.NoError(t, err)
 		policyClient := policy.NewPolicyController(informerCollection, nil, broker)
 
-		err = NewValidatingWebhook(webhook.Name, testNamespace, testVersion, testMeshName, enableReconciler, validateTrafficTarget, port, certManager, kube, policyClient, nil)
+		err = NewValidatingWebhook(context.Background(), webhook.Name, testNamespace, testVersion, testMeshName, enableReconciler, validateTrafficTarget, certManager, kube, policyClient)
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/validator/server_test.go]: done

File: pkg/webhook/server.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	}
+	return s, nil
+}
+
+// Run actually starts the server. It blocks until the passed in context is done.
+func (s *Server) Run(ctx context.Context) {
+	log.Info().Msgf("Starting conversion webhook server on: %s", s.server.Addr)
+	go func() {
+		err := s.server.ListenAndServeTLS("", "") // err is always non-nil
+		log.Error().Err(err).Msg("crd-converter webhook HTTP server failed to start")
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/webhook/server.go]: done

File: charts/osm/templates/osm-bootstrap-deployment.yaml [hunk 8]
```diff
@@ -74,8 +74,6 @@ spec:
               containerPort: 9443
             - name: "metrics"
               containerPort: 9091
-            - name: "health"
```
Comment (keithmattix) [inline charts/osm/templates/osm-bootstrap-deployment.yaml:77]: I get cutting down on unnecessary ports, but using the metrics port for healthchecks feels weird to me

File: cmd/osm-controller/osm-controller.go [hunk 9]
```diff
@@ -255,7 +255,7 @@ func main() {
 
 	clientset := extensionsClientset.NewForConfigOrDie(kubeConfig)
 
-	if err := validator.NewValidatingWebhook(validatorWebhookConfigName, osmNamespace, osmVersion, meshName, enableReconciler, validateTrafficTarget, constants.ValidatorWebhookPort, certManager, kubeClient, policyController, stop); err != nil {
+	if err := validator.NewValidatingWebhook(ctx, validatorWebhookConfigName, osmNamespace, osmVersion, meshName, enableReconciler, validateTrafficTarget, certManager, kubeClient, policyController); err != nil {
```
Comment (keithmattix) [inline cmd/osm-controller/osm-controller.go:258]: +1 for more context usage 😄

File: pkg/cli/verifier/control_plane.go [hunk 10]
```diff
@@ -48,9 +47,9 @@ func NewControlPlaneHealthVerifier(stdout io.Writer, stderr io.Writer, kubeClien
 		bootstrapProber: &podProber{
 			kubeClient: kubeClient,
 			restConfig: restConfig,
-			port:       crdconversion.HealthzPort,
+			port:       constants.CRDConversionWebhookPort,
 			path:       constants.WebhookHealthPath,
-			protocol:   constants.ProtocolHTTP,
+			protocol:   constants.ProtocolHTTPS,
```
Comment (keithmattix) [inline pkg/cli/verifier/control_plane.go:52]: I'm not 100% sure what this is being used for, but OSM doesn't support protocol HTTPS

File: pkg/webhook/server.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+	go func() {
+		err := s.server.ListenAndServeTLS("", "") // err is always non-nil
+		log.Error().Err(err).Msgf("%s webhook HTTP server shutdown")
+	}()
+
+	// Wait on exit signals
+	<-ctx.Done()
+
+	// Stop the servers
+	if err := s.server.Shutdown(ctx); err != nil {
```
Comment (keithmattix) [inline pkg/webhook/server.go:74]: You're passing a canceled context into `Shutdown` here; I'm not sure what the implications of that are.

File: pkg/webhook/server.go [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+		Addr:    fmt.Sprintf(":%d", port),
+		Handler: mux,
+		// #nosec G402
+		TLSConfig: &tls.Config{
+			GetCertificate: func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+				s.mu.Lock()
+				defer s.mu.Unlock()
+				return &s.cert, nil
+			},
+			MinVersion: tls.VersionTLS13,
```
Comment (keithmattix) [inline pkg/webhook/server.go:52]: Are we using this minVersion elsewhere? If any client doesn't support TLS 1.3, then the handshake will fail. I'd say 1.2 is probably safer.

File: pkg/cli/verifier/control_plane.go (hunk 10 above)
Comment (steeling) [reply pkg/cli/verifier/control_plane.go:52]: health check and metrics do

File: pkg/webhook/server.go (hunk 12 above)
Comment (steeling) [reply pkg/webhook/server.go:52]: This is what was used previously, I don't think it fits in the scope of this PR to make that change

File: pkg/webhook/server.go (hunk 11 above)
Comment (steeling) [reply pkg/webhook/server.go]: good catch! fixed

File: charts/osm/templates/osm-bootstrap-deployment.yaml (hunk 8 above)
Comment (keithmattix) [reply charts/osm/templates/osm-bootstrap-deployment.yaml:77]: Fair enough; yeah meta or admin makes sense to me, but I won't block this PR on it

**Review by nojnhuh — APPROVED** [review]

**Review by jaellio — APPROVED** [review]

**Pull request merged**